File: libs/backend/resume-python/devfolio_resume_python/service.py
"""
libs/backend/resume-python/src/service.py
ResumeService — the business logic layer.

Takes any IResumeRepository and uses it to fetch data.
All data methods accept a `resume` param that selects the variant.
get_tech_stack() reads live runtime info from the Python
process itself — no database needed for that part.
"""

from __future__ import annotations

import platform
import sys
from datetime import datetime, timezone

from .models import (
    DatabaseDialect,
    DatabaseInfo,
    Education,
    EnvironmentInfo,
    FrameworkInfo,
    OsInfo,
    PackageInfo,
    Profile,
    Project,
    RuntimeInfo,
    SkillCategory,
    TechStackInfo,
    WorkExperience,
)
from .repository import DEFAULT_RESUME, IResumeRepository


def _pkg_version(name: str) -> str:
    """Safely read an installed package version."""
    try:
        from importlib.metadata import version
        return version(name)
    except Exception:
        return 'unknown'


class ResumeService:

    def __init__(self, repository: IResumeRepository) -> None:
        self._repo = repository

    async def get_profile(self, resume: str = DEFAULT_RESUME) -> Profile:
        return await self._repo.find_profile(resume)

    async def get_work_experience(self, resume: str = DEFAULT_RESUME) -> list[WorkExperience]:
        return await self._repo.find_work_experience(resume)

    async def get_education(self, resume: str = DEFAULT_RESUME) -> list[Education]:
        return await self._repo.find_education(resume)

    async def get_skills(self, resume: str = DEFAULT_RESUME) -> list[SkillCategory]:
        return await self._repo.find_skills(resume)

    async def get_projects(self, resume: str = DEFAULT_RESUME) -> list[Project]:
        return await self._repo.find_projects(resume)

    async def get_tech_stack(self) -> TechStackInfo:
        """
        The phpinfo() of the stack — every value is live, nothing hardcoded.
        database.connected = False means we are in seed/fallback mode.
        """
        python_version   = sys.version.split()[0]
        implementation   = platform.python_implementation()
        fastapi_version  = _pkg_version('fastapi')
        uvicorn_version  = _pkg_version('uvicorn')
        pydantic_version = _pkg_version('pydantic')
        asyncpg_version  = _pkg_version('asyncpg')

        db_version   = await self._repo.get_database_version()
        db_name      = await self._repo.get_database_name()
        db_connected = 'seed mode' not in db_version.lower()

        dialect = DatabaseDialect.postgres
        if 'mysql' in db_name.lower():
            dialect = DatabaseDialect.mysql
        elif 'mssql' in db_name.lower() or 'sql server' in db_name.lower():
            dialect = DatabaseDialect.mssql
        elif 'mongo' in db_name.lower():
            dialect = DatabaseDialect.mongodb

        return TechStackInfo(
            generated_at=datetime.now(timezone.utc).isoformat(),
            runtime=RuntimeInfo(
                name='Python',
                version=python_version,
                implementation=implementation,
            ),
            framework=FrameworkInfo(
                name='FastAPI',
                version=fastapi_version,
                extra={
                    'uvicorn':  uvicorn_version,
                    'pydantic': pydantic_version,
                },
            ),
            database=DatabaseInfo(
                name=db_name,
                version=db_version,
                dialect=dialect,
                connected=db_connected,
            ),
            os=OsInfo(
                platform=platform.system(),
                release=platform.release(),
                architecture=platform.machine(),
            ),
            environment=EnvironmentInfo(
                name='development',
                timezone='UTC',
            ),
            packages=[
                PackageInfo(name='fastapi',  version=fastapi_version,  category='http'),
                PackageInfo(name='pydantic', version=pydantic_version, category='validation'),
                PackageInfo(name='uvicorn',  version=uvicorn_version,  category='server'),
                PackageInfo(name='asyncpg',  version=asyncpg_version,  category='database'),
            ],
        )

File: libs/backend/resume-ruby/lib/devfolio_resume_ruby/service.rb
# frozen_string_literal: true

require "rbconfig"
require "json"

module DevfolioResumeRuby
  class Service
    def initialize(repository)
      @repo = repository
    end

    def get_profile(resume:)
      @repo.get_profile(resume: resume)
    end

    def get_work_experience(resume:)
      @repo.get_work_experience(resume: resume)
    end

    def get_education(resume:)
      @repo.get_education(resume: resume)
    end

    def get_skills(resume:)
      @repo.get_skills(resume: resume)
    end

    def get_projects
      @repo.get_projects
    end

    def get_tech_stack
      db_version   = @repo.get_database_version
      db_name      = @repo.get_database_name
      db_connected = !db_version.downcase.include?("seed mode")

      TechStackInfo.new(
        generated_at: Time.now.utc.iso8601,
        runtime: RuntimeInfo.new(
          name:           "Ruby",
          version:        RUBY_VERSION,
          implementation: defined?(RUBY_ENGINE) ? RUBY_ENGINE : "ruby"
        ),
        framework: FrameworkInfo.new(
          name:    "Rails",
          version: Rails::VERSION::STRING,
          extra:   {
            "puma"   => gem_version("puma"),
            "pg"     => gem_version("pg")
          }
        ),
        database: DatabaseInfo.new(
          name:      db_name,
          version:   db_version,
          dialect:   "postgres",
          connected: db_connected
        ),
        os: OsInfo.new(
          platform:     RbConfig::CONFIG["host_os"],
          release:      `uname -r`.strip,
          architecture: RbConfig::CONFIG["host_cpu"]
        ),
        environment: EnvironmentInfo.new(
          name:     Rails.env,
          timezone: "UTC"
        ),
        packages: [
          PackageInfo.new(name: "rails",  version: Rails::VERSION::STRING, category: "framework"),
          PackageInfo.new(name: "puma",   version: gem_version("puma"),    category: "server"),
          PackageInfo.new(name: "pg",     version: gem_version("pg"),      category: "database"),
          PackageInfo.new(name: "rack",   version: gem_version("rack"),    category: "middleware")
        ]
      )
    end

    private

    def gem_version(name)
      Gem.loaded_specs[name]&.version&.to_s || "unknown"
    end
  end
end